File: C#/LocalSearchOptimization/Examples/Structures/Permutation/Shift.cs
﻿using System;
using System.Collections.Generic;
using LocalSearchOptimization.Components;

namespace LocalSearchOptimization.Examples.Structures.Permutation
{
    public class Shift : Operator
    {
        public Shift(int elementsNumber, double weight = 1) : base(weight)
        {
            for (int i = 0; i < elementsNumber; i++)
                for (int j = 0; j < elementsNumber; j++)
                    if (Math.Abs(i - j) >= 2) Configurations.Add(new TwoOperands(i, j, this));
        }

        public override ISolution Apply(ISolution solution, Configuration configuration)
        {
            IPermutation permutation = (IPermutation)solution;
            TwoOperands operands = (TwoOperands)configuration;

            List<int> shifted = new List<int>(permutation.Order);

            int shiftedItem = shifted[operands.First];
            shifted.RemoveAt(operands.First);
            if (operands.Second <= operands.First)
                shifted.Insert(operands.Second, shiftedItem);
            else
                shifted.Insert(operands.Second - 1, shiftedItem);

            return permutation.FetchPermutation(shifted, "shift");
        }
    }
}

File: java/src/localsearchoptimization/examples/structures/permutation/Shift.java
package localsearchoptimization.examples.structures.permutation;

import localsearchoptimization.components.Configuration;
import localsearchoptimization.components.Operator;
import localsearchoptimization.components.Solution;
import localsearchoptimization.examples.structures.TwoOperands;

public class Shift extends Operator {

    public Shift(int elementsNumber) {
        this(elementsNumber, 1);
    }

    public Shift(int elementsNumber, double weight) {
        super(weight);
        for (int i = 0; i < elementsNumber; i++)
            for (int j = 0; j < elementsNumber; j++)
                if (Math.abs(i - j) >= 2) configurations.add(new TwoOperands(i, j, this));
    }

    @Override
    public Solution apply(Solution solution, Configuration configuration) {
        Permutation permutation = (Permutation) solution;
        TwoOperands operands = (TwoOperands) configuration;

        int[] shifted = new int[permutation.getPermutation().length];

        shifted[operands.second] = permutation.getPermutation()[operands.first];

        for (int i = 0; i < Math.min(operands.first, operands.second); i++)
            shifted[i] = permutation.getPermutation()[i];

        for (int i = Math.max(operands.first, operands.second) + 1; i < shifted.length; i++)
            shifted[i] = permutation.getPermutation()[i];

        if (operands.first < operands.second)
            for (int i = operands.first; i < operands.second; i++)
                shifted[i] = permutation.getPermutation()[i + 1];
        else
            for (int i = operands.first; i > operands.second; i--)
                shifted[i] = permutation.getPermutation()[i - 1];

        return permutation.fetchPermutation(shifted, "shift");
    }
}